src/repr/src/adt/apd.rs

```diff
@@ -83,15 +156,207 @@ pub fn twos_complement_be_to_i128(input: &[u8]) -> Result<i128, anyhow::Error> {
+    assert!(d.exponent() >= 0);
+
+    let mut buf_cursor = 0;
+    while !d.is_zero() {
+        let mut w = d.clone();
+        // Take the remainder; this represents one of our "units" to take the coefficient of, i.e. d & i128::MAX
+        cx.rem(&mut w, D::unit_splitter());
+
+        // Take the `u128` version of the coefficient, which will always be what
+        // we want given that we adjusted negative values to have an unsigned
+        // integer representation.
+        let c = w.coefficient::<u128>().unwrap();
+
+        // Determine the width of the coefficient we want to take, i.e. the full
+        // coefficient or a part of it to fill the buffer.
+        let width = std::cmp::min(buf_cursor + S, D::TWOS_COMPLEMENT_BYTE_WIDTH);
+
+        // We're putting less significant bytes at index 0, which is little endian.
+        buf[buf_cursor..width].copy_from_slice(&c.to_le_bytes()[0..width - buf_cursor]);
+        // Advance cursor; ok that it will go past buffer on final + 1th iteration.
+        buf_cursor += S;
+
+        // Take the quotient to represent the next unit, i.e. d >> 128
+        cx.div_integer(&mut d, D::unit_splitter());
+    }
+
+    // Convert from little endian to big endian.
+    buf.reverse();
+}
+
+pub fn twos_complement_be_to_apd(input: &[u8]) -> Result<Apd, anyhow::Error> {
+    let mut cx = cx_datum();
+    Ok(if input.len() <= 17 {
+        let n = twos_complement_be_to_apd_inner::<ApdTwosComp, { APD_DATUM_WIDTH + 1 }>(input)?;
+        cx.to_width(n)
+    } else {
+        let n = twos_complement_be_to_apd_inner::<ApdAgg, APD_AGG_WIDTH>(input)?;
+        cx.to_width(n)
+    })
+}
+
+/// Parses a buffer of two's complement digits in big-endian order and converts
+/// them to [`Decimal<N>`].
+///
+/// Note that in cases where you want to use this value as a [`Datum::APD`], you
+/// will need to use `cx_datum()`'s `to_width` method.
+pub fn twos_complement_be_to_apd_inner<D: TwosCompDec<N>, const N: usize>(
+    input: &[u8],
+) -> Result<Decimal<N>, anyhow::Error> {
+    const S: usize = std::mem::size_of::<u128>();
+
+    // Only the first bytes should be parsed as signed, so take as few of those
+    // as will let us handle the rest as unsigned values.
+    let mut head = input.len() % S;
+    if head == 0 {
+        head = S;
+    }
+    let i = twos_complement_be_to_i128(&input[0..head]).unwrap();
+
+    let mut cx = D::twos_comp_cx();
+    let mut d = cx.from_i128(i);
+
+    for c in input[head..].chunks(S) {
+        assert_eq!(c.len(), S);
+        // essentially d << 128
+        cx.mul(&mut d, D::unit_splitter());
+        let i = twos_complement_be_to_u128(&c).unwrap();
+        let i = cx.from_u128(i);
+        cx.add(&mut d, &i);
+    }
+
+    if cx.status().inexact() {
+        bail!("value exceeds Apd")
+    } else if cx.status().any() {
+        bail!("unexpected status {:?}", cx.status());
+    }
+    Ok(d)
+}
+
+#[test]
+fn test_twos_complement_roundtrip() {
```

Comment:
btw, we'll want tests that explicitly check for the expected binary output as well since the output format is part of the API contract and shouldn't change. Testing roundtrip can sometimes hide bugs where an error is introduced in code shared by both the decoder and encoder such that they work properly together but produce output external systems don't consider valid.

I know the PR's only in draft, so apologies if you were already planning on adding these 🙂

Reply:
Oh yeah, definitely more tests incoming--this is just making sure that the logic actually did the things I was intending it to do.